### PR TITLE
Extend the definition of 'analyzer' in Semgrep rules

### PR DESCRIPTION
**Pull request by mjambon**
This PR extends the Semgrep rule schema in accordance with the work I did on the OCaml side. The PR is https://github.com/semgrep/semgrep-proprietary/pull/1641 and I'm going to push my work to it very soon.

This PR also adds JSON Schema validation tests. They're useful to check that the JSON Schema is valid, that it validates valid Semgrep rules, and rejects invalid Semgrep rules - without having to rely on another git repo to test this.

test plan: `make test`

This will run `check-jsonschema`. `make setup` will ask the user to install it if it's missing.

- [x] I ran `make setup && make` to update the generated code after editing a `.atd` file (TODO: have a CI check)
- [x] I made sure we're still backward compatible with old versions of the CLI.
      For example, the Semgrep backend need to still be able to *consume* data generated
	  by Semgrep 1.17.0.
      See https://atd.readthedocs.io/en/latest/atdgen-tutorial.html#smooth-protocol-upgrades

### CODE REVIEW

**Comment by aryx** [comment]
Can you also update rule_schema_v2.atd ?

File: rule_schema_v1.yaml [hunk 1]
```diff
@@ -482,13 +514,11 @@ $defs:
...
           analyzer:
-            type: string
-            title: Analyzer to use
+            $ref: "#/$defs/analyzer"
           metavariable:
             type: string
             title: Metavariable to analyze
-          options:
-            type: object
+          additionalProperties: false
```
Comment (aryx) [inline rule_schema_v1.yaml:521]: do we want that? Will this create problems down the line when we update the schema and add new fields?
or we can just rely on min-max version to skip the rule?

**Review by aryx — APPROVED** [review]

File: rule_schema_v1.yaml (hunk 1 above)
Comment (mjambon) [reply rule_schema_v1.yaml:521]: First, I made a mistake here. `additionalProperties:` should not be at the same level as `metavariable`.

Regardless, it's a good question. `additionalProperties` protects the user against spelling mistakes but comes with compatibility issues. I think I'll remove it.

Note that we could check additional fields for spelling mistakes and report an error if they're too close to a missing optional field name. We'd need to be careful to report e.g. `massage` but not `message2` if an optional `message` field isn't specified.

**Comment by mjambon** [comment]
> Can you also update rule_schema_v2.atd ?

I updated it. It's fairly complicated due an ATD adapter that can't be fully generic. It's explained in the comments along with a possible improvement in ATD that would simplify this.

I still have to check if this OCaml code compiles and works (I don't have the impression that we're testing it here).

File: rule_schema_v2.atd [hunk 2]
```diff
@@ -430,11 +430,19 @@ type metavariable_cond = {
 
 type mvar = string
```
Comment (aryx) [inline rule_schema_v2.atd:432]: this TODO comment is still valid? I can see an adapter below.

**Review by aryx — APPROVED** [review]

File: rule_schema_v2.atd (hunk 2 above)
Comment (mjambon) [reply rule_schema_v2.atd:432]: good catch, thx

**Pull request merged**